lib/os/json.c

```diff
@@ -390,6 +390,42 @@ static int arr_next(struct json_obj *json, struct json_token *value)
 	return element_token(value->type);
 }
 
+static void skip_field(struct json_obj *obj,
+		       struct json_obj_key_value *kv)
+{
+	int i = 1;
+
+	if (kv->value.type == JSON_TOK_ARRAY_START) {
+		while (arr_next(obj, &kv->value) >= 0) {
+			if (kv->value.type == JSON_TOK_ARRAY_END) {
+				i--;
+			} else if (kv->value.type == JSON_TOK_ARRAY_START) {
+				i++;
+			} else if (kv->value.type == JSON_TOK_OBJECT_START) {
+				skip_field(obj, kv);
```

Comment:
Given that this is part of OS library (maybe debatable the location?). I think we may not be allowed to use recursion due to MISRA C enforcement (#11425).

Maybe we can get away with not distinguishing between array and objects and just keeping a count (haven't tested as I'm away from my linux machine)
 
```
static void skip_field(struct json_obj *obj, struct json_obj_key_value *kv)
{
	int field_count = 1;
	if (kv->value.type == JSON_TOK_ARRAY_START || kv->value.type == JSON_TOK_ARRAY_END) {
		while (arr_next(obj, &kv->value) >= 0 && field_count > 0) {
			switch (descr->type) {
			case JSON_TOK_ARRAY_END:
			case JSON_TOK_OBJECT_END:
				field_count--;
				break;

			case JSON_TOK_OBJECT_START:
			case JSON_TOK_ARRAY_START:
				field_count++;
				break;
			}
		}
	}
}

```

Reply:
> Given that this is part of OS library (maybe debatable the location?). I think we may not be allowed to use recursion due to MISRA C enforcement (#11425).

Wasn't aware of that. The library already makes use of recursion (cf.` obj_parse()`, `arr_parse()` and `decode_value()`), so I followed that approach for the `skip_field()` function.

> Maybe we can get away with not distinguishing between array and objects and just keeping a count (haven't tested as I'm away from my linux machine)

Not sure how the suggested code skips nested objects. It will fail for sample 4. in https://github.com/zephyrproject-rtos/zephyr/issues/47988#issue-1309117307.

Reply:
> Wasn't aware of that. The library already makes use of recursion (cf. `obj_parse()`, `arr_parse()` and `decode_value()`), so I followed that approach for the `skip_field()` function.
> 

That's a good point, I've created #48102 to track it. I think this is due to when the json parser was added/written that it was not required. 

> Not sure how the suggested code skips nested objects. It will fail for sample 4. in https://github.com/zephyrproject-rtos/zephyr/issues/47988#issue-1309117307.

Apologies should have been:
```
static void skip_field(struct json_obj *obj, struct json_obj_key_value *kv)
{
	int field_count = 1;
	if (kv->value.type == JSON_TOK_ARRAY_START || kv->value.type == JSON_TOK_OBJECT_START) {
		while (field_count > 0 && arr_next(obj, &kv->value) >= 0) {
			switch (descr->type) {
			case JSON_TOK_ARRAY_END:
			case JSON_TOK_OBJECT_END:
				field_count--;
				break;

			case JSON_TOK_OBJECT_START:
			case JSON_TOK_ARRAY_START:
				field_count++;
				break;
			}
		}
	}
}
```
My though is it will continue to skip anything inside the initial array or object (by keeping track of start and end regardless of type - we don't care as we're skipping it- ) until the initial array or object end is reached. I'll do some testing later today or tomorrow.

Reply:
I followed your advice to get rid of the recursive calls of `skip_field()`.